File: experiments-local/experiment-mu-kimi3-brazil/runs/agent=opencode_language=typescript_model=openrouter/moonshotai/kimi-k3_tooling=none/rep1/src/lib/dataset.ts
/**
 * Dataset loader: reads the six Kaggle CSV files, normalizes every row
 * into canonical Match / Player records and merges duplicate matches
 * that appear in more than one file (datasets overlap heavily between
 * 2014-2022).
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { parse } from "csv-parse/sync";
import {
  Competition,
  Match,
  MatchStats,
  Player,
  Team,
} from "./types.js";
import { parseDateTime } from "./dates.js";
import { TeamRegistry } from "./teams.js";
import { normalizeText } from "./text.js";

export const DATA_FILES = {
  brasileirao: "Brasileirao_Matches.csv",
  copaDoBrasil: "Brazilian_Cup_Matches.csv",
  libertadores: "Libertadores_Matches.csv",
  extended: "BR-Football-Dataset.csv",
  historical: "novo_campeonato_brasileiro.csv",
  fifa: "fifa_data.csv",
} as const;

/** Locate the data/kaggle directory (env override, cwd, or module-relative). */
export function resolveDataDir(): string {
  if (process.env.DATA_DIR && fs.existsSync(process.env.DATA_DIR)) {
    return process.env.DATA_DIR;
  }
  const here = path.dirname(fileURLToPath(import.meta.url));
  const candidates = [
    path.resolve(process.cwd(), "data", "kaggle"),
    path.resolve(here, "..", "..", "data", "kaggle"), // from src/lib or dist/lib
    path.resolve(here, "..", "..", "..", "data", "kaggle"),
  ];
  for (const c of candidates) {
    if (fs.existsSync(path.join(c, DATA_FILES.brasileirao))) return c;
  }
  throw new Error(
    `Could not locate data/kaggle directory. Set DATA_DIR. Tried: ${candidates.join(", ")}`,
  );
}

function readCsv(dataDir: string, file: string): Record<string, string>[] {
  const raw = fs.readFileSync(path.join(dataDir, file), "utf-8");
  return parse(raw, {
    columns: true,
    skip_empty_lines: true,
    relax_column_count: true,
    trim: true,
    bom: true,
  }) as Record<string, string>[];
}

function parseGoals(raw: string | undefined | null): number | null {
  if (raw === undefined || raw === null) return null;
  const s = String(raw).trim();
  if (s === "" || s.toUpperCase() === "NA") return null;
  const n = Number(s);
  return Number.isFinite(n) ? Math.round(n) : null;
}

function parseIntOrNull(raw: string | undefined | null): number | null {
  if (raw === undefined || raw === null) return null;
  const s = String(raw).trim();
  if (s === "" || s.toUpperCase() === "NA") return null;
  const n = Number(s);
  return Number.isFinite(n) ? Math.trunc(n) : null;
}

function parseSeason(raw: string | undefined | null): number | null {
  return parseIntOrNull(raw);
}

function num(raw: string | undefined | null): number | null {
  if (raw === undefined || raw === null || String(raw).trim() === "") return null;
  const n = Number(String(raw).trim());
  return Number.isFinite(n) ? n : null;
}

/** Parse a float-ish score like "1.0" into half-time goals. */
function halfTime(raw: string | undefined | null): number | null {
  return num(raw);
}

export interface Dataset {
  matches: Match[];
  players: Player[];
  teams: TeamRegistry;
  /** Per-file row counts actually loaded. */
  loadReport: Record<string, number>;
}

interface MatchDraft {
  competition: Competition;
  season: number | null;
  round: string | null;
  date: string | null;
  time: string | null;
  home: Team;
  away: Team;
  homeGoals: number | null;
  awayGoals: number | null;
  stadium: string | null;
  stats: MatchStats | null;
  source: string;
}

function dedupeKey(d: MatchDraft): string {
  return [
    d.competition,
    d.date ?? "unknown-date",
    d.home.key,
    d.away.key,
  ].join("|");
}

/**
 * Known mislabeled rows in the source files, skipped at load time.
 * BR-Football-Dataset.csv row "2016-01-30, Brasilia FC 1-1 CA Taguatinga"
 * is labeled "Serie A" — neither club has ever played Série A (it is a
 * 2016 Copa do Brasil qualifying-round tie).
 */
const SOURCE_DATA_EXCLUSIONS: { date: string; home: string; away: string }[] = [
  { date: "2016-01-30", home: "Brasilia FC", away: "CA Taguatinga" },
];

function isExcluded(dt: ReturnType<typeof parseDateTime>, home: string, away: string): boolean {
  return SOURCE_DATA_EXCLUSIONS.some(
    (e) => e.date === dt?.date && e.home === home.trim() && e.away === away.trim(),
  );
}

function isoDayNumber(date: string): number {
  return Date.parse(`${date}T00:00:00Z`) / 86_400_000;
}

function mergeMatchInto(target: Match, d: MatchDraft | Match, source: string): void {
  if (!target.sources.includes(source)) target.sources.push(source);
  target.round ??= d.round;
  target.stadium ??= d.stadium;
  target.time ??= d.time;
  target.season ??= d.season;
  target.homeGoals ??= d.homeGoals;
  target.awayGoals ??= d.awayGoals;
  target.stats ??= d.stats;
}

/**
 * Second dedupe pass: datasets occasionally record the same fixture one or
 * two days apart (reschedules, timezone cuts), and the extended dataset
 * labels matches by calendar year — so the COVID-delayed 2020 season,
 * played into early 2021, carries a different "season" than the canonical
 * files. Group by (competition, home, away) only — a real fixture occurs
 * at most once per season with the same home/away assignment, and distinct
 * seasons are always far more than two days apart — then merge records
 * whose dates differ by at most `maxDaysApart`. Records loaded first
 * (canonical files with explicit season columns) win, so merged fixtures
 * keep the correct season.
 */
function fuzzyMergeFixtures(matches: Map<string, Match>, maxDaysApart = 2): Map<string, Match> {
  const byFixture = new Map<string, Match[]>();
  for (const m of matches.values()) {
    if (m.date === null) continue;
    const key = [m.competition, m.homeTeam.key, m.awayTeam.key].join("|");
    if (!byFixture.has(key)) byFixture.set(key, []);
    byFixture.get(key)!.push(m);
  }
  const dropped = new Set<string>();
  for (const group of byFixture.values()) {
    if (group.length < 2) continue;
    // Multi-anchor clustering in load order (canonical sources first): a
    // record joins the first cluster within `maxDaysApart`, otherwise it
    // starts a new cluster (e.g. the same fixture in a different season).
    const anchors: Match[] = [];
    for (const m of group) {
      let target: Match | null = null;
      for (const a of anchors) {
        if (Math.abs(isoDayNumber(m.date!) - isoDayNumber(a.date!)) <= maxDaysApart) {
          target = a;
          break;
        }
      }
      if (target) {
        for (const s of m.sources) mergeMatchInto(target, m, s);
        dropped.add(m.id);
      } else {
        anchors.push(m);
      }
    }
  }

  // Stale-schedule pass: postponed matches appear as an unplayed record
  // (scheduled date, NA goals) in one file and as the actually-played
  // record (later date) in another. A fixture occurs at most once per
  // season per home/away pair, so an unplayed record that has a played
  // counterpart in the same season (or within 60 days) is a stale
  // schedule entry — fold it into the played record. Matches that were
  // genuinely never played (e.g. Chapecoense x Atlético-MG, 2016 round 38)
  // have no played counterpart and survive untouched.
  for (const group of byFixture.values()) {
    const playedRecords = group.filter(
      (m) => !dropped.has(m.id) && m.homeGoals !== null && m.awayGoals !== null,
    );
    if (playedRecords.length === 0) continue;
    for (const m of group) {
      if (dropped.has(m.id) || (m.homeGoals !== null && m.awayGoals !== null)) continue;
      const target = playedRecords.find(
        (p) =>
          (m.season !== null && p.season === m.season) ||
          Math.abs(isoDayNumber(p.date!) - isoDayNumber(m.date!)) <= 60,
      );
      if (target) {
        for (const s of m.sources) mergeMatchInto(target, m, s);
        dropped.add(m.id);
      }
    }
  }
  if (dropped.size === 0) return matches;
  const result = new Map<string, Match>();
  for (const [key, m] of matches) {
    if (!dropped.has(m.id)) result.set(key, m);
  }
  return result;
}

/** Load every CSV file and return the unified, deduplicated dataset. */
export function loadDataset(dataDir?: string): Dataset {
  const dir = dataDir ?? resolveDataDir();
  const teams = new TeamRegistry();
  const matches = new Map<string, Match>();
  const loadReport: Record<string, number> = {};
  let counter = 0;

  const addMatch = (d: MatchDraft): void => {
    const key = dedupeKey(d);
    const existing = matches.get(key);
    if (existing) {
      // Merge: fill gaps and record provenance.
      if (!existing.sources.includes(d.source)) existing.sources.push(d.source);
      existing.round ??= d.round;
      existing.stadium ??= d.stadium;
      existing.time ??= d.time;
      existing.season ??= d.season;
      existing.homeGoals ??= d.homeGoals;
      existing.awayGoals ??= d.awayGoals;
      existing.stats ??= d.stats;
      return;
    }
    matches.set(key, {
      id: `m${++counter}`,
      competition: d.competition,
      season: d.season,
      round: d.round,
      date: d.date,
      time: d.time,
      homeTeam: d.home,
      awayTeam: d.away,
      homeGoals: d.homeGoals,
      awayGoals: d.awayGoals,
      stadium: d.stadium,
      stats: d.stats,
      sources: [d.source],
    });
  };

  // 1. Brasileirão Série A (2012-2022), canonical round info.
  const brasileirao = readCsv(dir, DATA_FILES.brasileirao);
  for (const r of brasileirao) {
    const dt = parseDateTime(r.datetime);
    addMatch({
      competition: Competition.BrasileiraoSerieA,
      season: parseSeason(r.season),
      round: r.round?.trim() || null,
      date: dt?.date ?? null,
      time: dt?.time ?? null,
      home: teams.register(r.home_team, r.home_team_state),
      away: teams.register(r.away_team, r.away_team_state),
      homeGoals: parseGoals(r.home_goal),
      awayGoals: parseGoals(r.away_goal),
      stadium: null,
      stats: null,
      source: DATA_FILES.brasileirao,
    });
  }
  loadReport[DATA_FILES.brasileirao] = brasileirao.length;

  // 2. Copa do Brasil (2012-2021).
  const cup = readCsv(dir, DATA_FILES.copaDoBrasil);
  for (const r of cup) {
    const dt = parseDateTime(r.datetime);
    addMatch({
      competition: Competition.CopaDoBrasil,
      season: parseSeason(r.season),
      round: r.round?.trim() || null,
      date: dt?.date ?? null,
      time: dt?.time ?? null,
      home: teams.register(r.home_team),
      away: teams.register(r.away_team),
      homeGoals: parseGoals(r.home_goal),
      awayGoals: parseGoals(r.away_goal),
      stadium: null,
      stats: null,
      source: DATA_FILES.copaDoBrasil,
    });
  }
  loadReport[DATA_FILES.copaDoBrasil] = cup.length;

  // 3. Copa Libertadores (2013-2022).
  const libertadores = readCsv(dir, DATA_FILES.libertadores);
  for (const r of libertadores) {
    const dt = parseDateTime(r.datetime);
    addMatch({
      competition: Competition.Libertadores,
      season: parseSeason(r.season),
      round: r.stage?.trim() || null,
      date: dt?.date ?? null,
      time: dt?.time ?? null,
      home: teams.register(r.home_team),
      away: teams.register(r.away_team),
      homeGoals: parseGoals(r.home_goal),
      awayGoals: parseGoals(r.away_goal),
      stadium: null,
      stats: null,
      source: DATA_FILES.libertadores,
    });
  }
  loadReport[DATA_FILES.libertadores] = libertadores.length;

  // 4. Historical Brasileirão (2003-2019), Portuguese columns, DD/MM/YYYY.
  const historical = readCsv(dir, DATA_FILES.historical);
  for (const r of historical) {
    const dt = parseDateTime(r.Data);
    addMatch({
      competition: Competition.BrasileiraoSerieA,
      season: parseSeason(r.Ano) ?? (dt ? Number(dt.date.slice(0, 4)) : null),
      round: r.Rodada?.trim() || null,
      date: dt?.date ?? null,
      time: dt?.time ?? null,
      home: teams.register(r.Equipe_mandante, r.Mandante_UF),
      away: teams.register(r.Equipe_visitante, r.Visitante_UF),
      homeGoals: parseGoals(r.Gols_mandante),
      awayGoals: parseGoals(r.Gols_visitante),
      stadium: r.Arena?.trim() || null,
      stats: null,
      source: DATA_FILES.historical,
    });
  }
  loadReport[DATA_FILES.historical] = historical.length;

  // 5. Extended dataset (2014-2023): Série A/B/C + Copa do Brasil w/ stats.
  const extended = readCsv(dir, DATA_FILES.extended);
  const tournamentMap: Record<string, Competition> = {
    "serie a": Competition.BrasileiraoSerieA,
    "serie b": Competition.SerieB,
    "serie c": Competition.SerieC,
    "copa do brasil": Competition.CopaDoBrasil,
  };
  for (const r of extended) {
    const competition = tournamentMap[normalizeText(r.tournament ?? "")];
    if (!competition) continue;
    const dt = parseDateTime(r.date);
    if (isExcluded(dt, r.home, r.away)) continue;
    const htHome = halfTime(r.ht_diff);
    const stats: MatchStats = {
      homeCorners: num(r.home_corner),
      awayCorners: num(r.away_corner),
      homeShots: num(r.home_shots),
      awayShots: num(r.away_shots),
      homeAttacks: num(r.home_attack),
      awayAttacks: num(r.away_attack),
      // ht_diff/at_diff are goal differences; recover absolute HT score when possible.
      halfTimeHomeGoals: null,
      halfTimeAwayGoals: null,
    };
    const hg = parseGoals(r.home_goal);
    const ag = parseGoals(r.away_goal);
    if (htHome !== null && hg !== null && ag !== null) {
      // ht_diff = HT home - HT away; we cannot recover absolutes reliably, skip.
    }
    addMatch({
      competition,
      season: dt ? Number(dt.date.slice(0, 4)) : null,
      round: null,
      date: dt?.date ?? null,
      time: r.time?.trim() ? r.time.trim().slice(0, 5) : (dt?.time ?? null),
      home: teams.register(r.home),
      away: teams.register(r.away),
      homeGoals: hg,
      awayGoals: ag,
      stadium: null,
      stats,
      source: DATA_FILES.extended,
    });
  }
  loadReport[DATA_FILES.extended] = extended.length;

  // 6. FIFA players.
  const fifa = readCsv(dir, DATA_FILES.fifa);
  const skillColumns = [
    "Crossing", "Finishing", "HeadingAccuracy", "ShortPassing", "Volleys",
    "Dribbling", "Curve", "FKAccuracy", "LongPassing", "BallControl",
    "Acceleration", "SprintSpeed", "Agility", "Reactions", "Balance",
    "ShotPower", "Jumping", "Stamina", "Strength", "LongShots",
    "Aggression", "Interceptions", "Positioning", "Vision", "Penalties",
    "Composure", "Marking", "StandingTackle", "SlidingTackle",
    "GKDiving", "GKHandling", "GKKicking", "GKPositioning", "GKReflexes",
  ];
  const players: Player[] = [];
  for (const r of fifa) {
    const club = r.Club?.trim() || null;
    let teamKey: string | null = null;
    if (club) {
      const res = teams.resolve(club);
      if (res.team) teamKey = res.team.key;
    }
    const skills: Record<string, number | null> = {};
    for (const c of skillColumns) skills[c] = num(r[c]);
    players.push({
      id: parseIntOrNull(r.ID) ?? players.length,
      name: r.Name ?? "",
      age: parseIntOrNull(r.Age),
      nationality: r.Nationality ?? "",
      overall: parseIntOrNull(r.Overall),
      potential: parseIntOrNull(r.Potential),
      club,
      position: r.Position?.trim() || null,
      jerseyNumber: parseIntOrNull(r["Jersey Number"]),
      height: r.Height?.trim() || null,
      weight: r.Weight?.trim() || null,
      preferredFoot: r["Preferred Foot"]?.trim() || null,
      skills,
      teamKey,
    });
  }
  loadReport[DATA_FILES.fifa] = players.length;

  const deduped = fuzzyMergeFixtures(matches);
  const matchList = [...deduped.values()];
  matchList.sort((a, b) => (a.date ?? "").localeCompare(b.date ?? ""));

  return { matches: matchList, players, teams, loadReport };
}

File: experiments-local/experiment-mu-glm53-brazil/runs/agent=opencode_language=python_model=openrouter/z-ai/glm-5.3_tooling=none/rep3/brazilian_soccer_mcp/dataset.py
"""
Dataset loading and assembly for the Brazilian Soccer MCP server.

Context block
-------------
Why:
    Five match CSVs and one player CSV overlap and disagree: the same
    Brasileirão fixture appears in up to three files with different team
    spellings and occasionally different dates, and BR-Football-Dataset
    even contains duplicated rows.  Naive concatenation would double-count
    matches, skew standings and break head-to-head records, so the data is
    assembled into a *canonical* match index before any query runs.

What:
    * ``load_matches`` / ``load_players`` - parse the six CSVs into typed
      records (``Match`` / ``Player``), tolerating every observed quirk:
      float-formatted goals ("1.0"), "NA" scores, DD/MM/YYYY vs ISO dates,
      state-suffixed team names, UTF-8 accents.
    * ``Dataset`` - the assembled knowledge graph:
        - ``registry``  : every team spelling -> ``Club`` entity,
        - ``raw_matches``: all parsed rows from the five match files,
        - ``matches``    : canonical index - for each (competition, season)
          only the highest-priority source is kept (dedicated files first,
          then novo 2003-2011 history, then BR-Football for 2022-2023),
          which removes cross-file double counting,
        - ``players``    : the FIFA database, joined to clubs by registry.
    * ``load_dataset`` - builds and caches the ``Dataset`` (loading is
      ~1s; queries then run in-memory, meeting the TASK.md performance
      targets of <2s simple / <5s aggregate lookups).

Test:
    BDD GWT scenarios in ``tests/test_dataset.py``: file coverage counts,
    canonical-source selection, absence of duplicate fixtures per season,
    and club resolution of every canonical match.

Spec references:
    TASK.md "Provided Data" (all six files must be loadable and queryable),
    "Success Criteria" -> "Data Coverage" ("All 6 CSV files are loadable
    and queryable", "Cross-file queries work").
"""

from __future__ import annotations

import csv
import threading
from dataclasses import dataclass, field
from pathlib import Path

from .models import Match, MatchStats, Player
from .normalize import (
    TeamName,
    parse_date,
    parse_int,
    resolve_competition,
)
from .registry import ClubRegistry

DATA_DIR_DEFAULT = Path(__file__).resolve().parent.parent / "data" / "kaggle"

#: Lower number = higher preference when several sources cover the same
#: (competition, season).  Dedicated competition files beat the historical
#: novo file (overlaps 2012-2019 with shifted dates) which beats the
#: BR-Football aggregate (incomplete early seasons, duplicated 2021 rows).
SOURCE_PRIORITY: dict[str, int] = {
    "Brasileirao_Matches": 0,
    "Brazilian_Cup_Matches": 0,
    "Libertadores_Matches": 0,
    "novo_campeonato_brasileiro": 1,
    "BR-Football-Dataset": 2,
}

#: FIFA skill rating columns exposed on each player record.
_SKILL_COLUMNS = (
    "Crossing",
    "Finishing",
    "HeadingAccuracy",
    "ShortPassing",
    "Volleys",
    "Dribbling",
    "Curve",
    "FKAccuracy",
    "LongPassing",
    "BallControl",
    "Acceleration",
    "SprintSpeed",
    "Agility",
    "Reactions",
    "Balance",
    "ShotPower",
    "Jumping",
    "Stamina",
    "Strength",
    "LongShots",
    "Aggression",
    "Interceptions",
    "Positioning",
    "Vision",
    "Penalties",
    "Composure",
    "StandingTackle",
    "SlidingTackle",
)


@dataclass
class Dataset:
    """Assembled, deduplicated soccer knowledge base."""

    registry: ClubRegistry = field(default_factory=ClubRegistry)
    raw_matches: list[Match] = field(default_factory=list)
    matches: list[Match] = field(default_factory=list)  # canonical index
    players: list[Player] = field(default_factory=list)
    #: (competition, season) -> {"source": str, "matches": int}
    season_sources: dict[tuple[str, int | None], dict] = field(default_factory=dict)
    #: per-file load statistics
    file_stats: dict[str, dict] = field(default_factory=dict)

    # -- club helpers ----------------------------------------------------

    def club_display(self, team: TeamName) -> str:
        club = self.registry.get(team)
        return club.display if club else (team.base or "?")

    def club_id(self, team: TeamName) -> str:
        club = self.registry.get(team)
        return club.id if club else team.key

    def matches_for_club(self, club_id: str, *, canonical_only: bool = True) -> list[Match]:
        pool = self.matches if canonical_only else self.raw_matches
        return [m for m in pool if (m._home_club == club_id) or (m._away_club == club_id)]


# --------------------------------------------------------------------------
# CSV loaders
# --------------------------------------------------------------------------


def _iter_csv(path: Path) -> list[dict]:
    with path.open(encoding="utf-8", newline="") as handle:
        return list(csv.DictReader(handle))


def _clean_raw_name(value: str) -> str:
    return (value or "").strip()


def _make_match(
    *,
    source: str,
    date: object,
    home_raw: str,
    away_raw: str,
    home_goals: object,
    away_goals: object,
    competition: str,
    season: object,
    round_label: str | None = None,
    stage: str | None = None,
    stadium: str | None = None,
    stats: MatchStats | None = None,
) -> Match:
    from .normalize import normalize_team

    home_raw = _clean_raw_name(home_raw)
    away_raw = _clean_raw_name(away_raw)
    return Match(
        date=parse_date(date),
        home_name=home_raw,
        away_name=away_raw,
        home_team=normalize_team(home_raw),
        away_team=normalize_team(away_raw),
        home_goals=parse_int(home_goals),
        away_goals=parse_int(away_goals),
        competition=competition,
        season=parse_int(season),
        round=round_label,
        stage=stage,
        source=source,
        stadium=stadium,
        stats=stats,
    )


def _load_brasileirao(data_dir: Path) -> list[Match]:
    rows = _iter_csv(data_dir / "Brasileirao_Matches.csv")
    return [
        _make_match(
            source="Brasileirao_Matches",
            date=r.get("datetime"),
            home_raw=r.get("home_team", ""),
            away_raw=r.get("away_team", ""),
            home_goals=r.get("home_goal"),
            away_goals=r.get("away_goal"),
            competition="Brasileirão Serie A",
            season=r.get("season"),
            round_label=str(r.get("round") or "") or None,
        )
        for r in rows
    ]


def _load_cup(data_dir: Path) -> list[Match]:
    rows = _iter_csv(data_dir / "Brazilian_Cup_Matches.csv")
    return [
        _make_match(
            source="Brazilian_Cup_Matches",
            date=r.get("datetime"),
            home_raw=r.get("home_team", ""),
            away_raw=r.get("away_team", ""),
            home_goals=r.get("home_goal"),
            away_goals=r.get("away_goal"),
            competition="Copa do Brasil",
            season=r.get("season"),
            round_label=str(r.get("round") or "") or None,
        )
        for r in rows
    ]


def _load_libertadores(data_dir: Path) -> list[Match]:
    rows = _iter_csv(data_dir / "Libertadores_Matches.csv")
    return [
        _make_match(
            source="Libertadores_Matches",
            date=r.get("datetime"),
            home_raw=r.get("home_team", ""),
            away_raw=r.get("away_team", ""),
            home_goals=r.get("home_goal"),
            away_goals=r.get("away_goal"),
            competition="Copa Libertadores",
            season=r.get("season"),
            stage=(r.get("stage") or "").strip() or None,
        )
        for r in rows
    ]


def _load_novo(data_dir: Path) -> list[Match]:
    rows = _iter_csv(data_dir / "novo_campeonato_brasileiro.csv")
    out: list[Match] = []
    for r in rows:
        out.append(
            _make_match(
                source="novo_campeonato_brasileiro",
                date=r.get("Data"),
                home_raw=r.get("Equipe_mandante", ""),
                away_raw=r.get("Equipe_visitante", ""),
                home_goals=r.get("Gols_mandante"),
                away_goals=r.get("Gols_visitante"),
                competition="Brasileirão Serie A",
                season=r.get("Ano"),
                round_label=str(r.get("Rodada") or "") or None,
                stadium=(r.get("Arena") or "").strip() or None,
            )
        )
    return out


def _load_br_football(data_dir: Path) -> list[Match]:
    rows = _iter_csv(data_dir / "BR-Football-Dataset.csv")
    out: list[Match] = []
    for r in rows:
        tournament = (r.get("tournament") or "").strip()
        competition = resolve_competition(tournament) or tournament
        stats = MatchStats(
            home_corners=parse_int(r.get("home_corner")),
            away_corners=parse_int(r.get("away_corner")),
            home_shots=parse_int(r.get("home_shots")),
            away_shots=parse_int(r.get("away_shots")),
            home_attacks=parse_int(r.get("home_attack")),
            away_attacks=parse_int(r.get("away_attack")),
            home_ht_result=(r.get("ht_result") or "").strip() or None,
            away_ht_result=(r.get("at_result") or "").strip() or None,
            kickoff_time=(r.get("time") or "").strip() or None,
        )
        out.append(
            _make_match(
                source="BR-Football-Dataset",
                date=r.get("date"),
                home_raw=r.get("home", ""),
                away_raw=r.get("away", ""),
                home_goals=r.get("home_goal"),
                away_goals=r.get("away_goal"),
                competition=competition,
                season=(r.get("date") or "")[:4] or None,
                stats=stats,
            )
        )
    return out


def load_matches(data_dir: Path = DATA_DIR_DEFAULT) -> tuple[list[Match], dict[str, int]]:
    """Load all five match CSVs; returns (raw matches, per-file row counts)."""
    loaders = {
        "Brasileirao_Matches": _load_brasileirao,
        "Brazilian_Cup_Matches": _load_cup,
        "Libertadores_Matches": _load_libertadores,
        "novo_campeonato_brasileiro": _load_novo,
        "BR-Football-Dataset": _load_br_football,
    }
    raw: list[Match] = []
    counts: dict[str, int] = {}
    for name, loader in loaders.items():
        matches = loader(data_dir)
        counts[name] = len(matches)
        raw.extend(matches)
    return raw, counts


def load_players(data_dir: Path = DATA_DIR_DEFAULT) -> list[Player]:
    """Load the FIFA player database."""
    rows = _iter_csv(data_dir / "fifa_data.csv")
    players: list[Player] = []
    for r in rows:
        name = (r.get("Name") or "").strip()
        if not name:
            continue
        overall = parse_int(r.get("Overall")) or 0
        potential = parse_int(r.get("Potential")) or overall
        skills: dict[str, int] = {}
        for col in _SKILL_COLUMNS:
            value = parse_int(r.get(col))
            if value is not None:
                skills[col.lower()] = value
        players.append(
            Player(
                player_id=parse_int(r.get("ID")) or 0,
                name=name,
                age=parse_int(r.get("Age")),
                nationality=(r.get("Nationality") or "").strip(),
                overall=overall,
                potential=potential,
                club=(r.get("Club") or "").strip(),
                position=(r.get("Position") or "").strip() or None,
                jersey_number=parse_int(r.get("Jersey Number")),
                preferred_foot=(r.get("Preferred Foot") or "").strip() or None,
                value=(r.get("Value") or "").strip() or None,
                wage=(r.get("Wage") or "").strip() or None,
                height=(r.get("Height") or "").strip() or None,
                weight=(r.get("Weight") or "").strip() or None,
                skills=skills,
            )
        )
    return players


# --------------------------------------------------------------------------
# Assembly
# --------------------------------------------------------------------------


def _build_canonical(
    raw: list[Match], dataset: Dataset
) -> tuple[list[Match], dict[tuple[str, int | None], dict]]:
    """Pick one authoritative source per (competition, season) and dedup.

    Guarantees a fixture (date, home club, away club) appears at most once
    within a competition-season even if the chosen source repeats it.
    """
    groups: dict[tuple[str, int | None], list[Match]] = {}
    for m in raw:
        if not m.home_team.base and not m.away_team.base:
            continue  # unparseable row (no teams at all)
        groups.setdefault((m.competition, m.season), []).append(m)

    canonical: list[Match] = []
    season_sources: dict[tuple[str, int | None], dict] = {}
    for key, candidates in sorted(groups.items(), key=lambda kv: (kv[0][0], kv[0][1] or 0)):
        best_source = min(
            {m.source for m in candidates},
            key=lambda s: SOURCE_PRIORITY.get(s, 99),
        )
        chosen = [m for m in candidates if m.source == best_source]
        seen: set[tuple] = set()
        deduped: list[Match] = []
        for m in sorted(chosen, key=lambda x: (x.date is None, x.date or "")):
            ident = (m.date, m._home_club, m._away_club)
            if ident in seen:
                continue
            seen.add(ident)
            deduped.append(m)
        season_sources[key] = {"source": best_source, "matches": len(deduped)}
        canonical.extend(deduped)

    canonical.sort(key=lambda m: (m.date is None, m.date or "", m.competition))
    return canonical, season_sources


def _attach_extended_stats(dataset: Dataset) -> None:
    """Join BR-Football extended stats onto canonical matches.

    BR-Football is the only source with corners/shots/attacks and half-time
    labels; when it is not the authoritative source for a season the stats
    are attached to the canonical record by (date, home club, away club).
    """
    by_key: dict[tuple, MatchStats] = {}
    for m in dataset.raw_matches:
        if m.source != "BR-Football-Dataset" or m.stats is None:
            continue
        if not m.date:
            continue
        key = (m.date, m._home_club, m._away_club)
        if key not in by_key:
            by_key[key] = m.stats
    attached = 0
    for m in dataset.matches:
        if m.stats is not None or not m.date:
            continue
        stats = by_key.get((m.date, m._home_club, m._away_club))
        if stats is not None:
            m.stats = stats
            attached += 1
    dataset.file_stats["extended_stats_attached"] = {"matches": attached}


def build_dataset(data_dir: Path = DATA_DIR_DEFAULT) -> Dataset:
    """Load everything and assemble the knowledge base."""
    dataset = Dataset()
    raw, counts = load_matches(data_dir)
    dataset.file_stats.update({name: {"rows": n} for name, n in counts.items()})
    dataset.raw_matches = raw
    players = load_players(data_dir)
    dataset.players = players
    dataset.file_stats["fifa_data"] = {"rows": len(players)}

    # 1. Register every team spelling from the match files.
    for m in raw:
        if m.home_team.base:
            dataset.registry.register(m.home_name)
        if m.away_team.base:
            dataset.registry.register(m.away_name)

    # 2. Count match appearances per club (drives merge dominance).
    for m in raw:
        if m.home_team.base:
            dataset.registry.add_match_count(m.home_team)
        if m.away_team.base:
            dataset.registry.add_match_count(m.away_team)

    # 3. Register FIFA club strings so they join the same graph.
    for p in players:
        if p.club:
            dataset.registry.register(p.club)

    # 4. Merge state-less variants into dominant state-qualified clubs.
    dataset.registry.finalize()

    # 5. Resolve club ids on every raw match (used by canonical build).
    for m in raw:
        m._home_club = dataset.club_id(m.home_team)
        m._away_club = dataset.club_id(m.away_team)

    # 6. Canonical index + extended-stat join.
    dataset.matches, dataset.season_sources = _build_canonical(raw, dataset)
    _attach_extended_stats(dataset)

    # 7. Attach players to clubs and set display names on canonical matches.
    for p in players:
        from .normalize import normalize_team

        if p.club:
            p._club_team = normalize_team(p.club)
            club = dataset.registry.get(p._club_team)
            if club is not None:
                p._club_id = club.id
                club.player_count += 1
    for m in dataset.matches:
        m.home_name = dataset.club_display(m.home_team)
        m.away_name = dataset.club_display(m.away_team)

    return dataset


# --------------------------------------------------------------------------
# Cached singleton access
# --------------------------------------------------------------------------

_lock = threading.Lock()
_cache: dict[Path, Dataset] = {}


def load_dataset(data_dir: Path | str | None = None) -> Dataset:
    """Return the cached ``Dataset`` for ``data_dir`` (default: bundled data)."""
    path = Path(data_dir) if data_dir else DATA_DIR_DEFAULT
    path = path.resolve()
    with _lock:
        if path not in _cache:
            _cache[path] = build_dataset(path)
        return _cache[path]
